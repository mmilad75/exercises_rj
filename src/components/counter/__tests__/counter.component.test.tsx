import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Counter from "../counter.component";

describe("Counter component", () => {
    it("should render correctly", () => {
        const { asFragment } = render(<Counter />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("should display increment button", () => {
        render(<Counter />);
        const incrementButton = screen.getByText("Increment");

        expect(incrementButton).toBeInTheDocument();
    });

    it("should display decrement button", () => {
        render(<Counter />);
        const decrementButton = screen.getByText("Decrement");

        expect(decrementButton).toBeInTheDocument();
    });

    it("should display correct number of step options", () => {
        render(<Counter />);
        const options = screen.getAllByRole("option");

        expect(options.length).toEqual(3);
    });

    it("should set default counter correctly", () => {
        render(<Counter />);
        const counterText = screen.getByText("Counter: 0");

        expect(counterText).toBeInTheDocument();
    });

    it("should set default step correctly", () => {
        render(<Counter />);
        const firstOption = screen.getByRole<HTMLOptionElement>("option", {
            name: "1",
        });

        expect(firstOption.selected).toEqual(true);
    });

    it("should allow user to change step", () => {
        render(<Counter />);

        userEvent.selectOptions(
            screen.getByRole("combobox"),
            screen.getByRole("option", { name: "10" })
        );

        const option = screen.getByRole<HTMLOptionElement>("option", {
            name: "10",
        });

        expect(option.selected).toBe(true);
    });

    it("should increment the counter by selected step", () => {
        render(<Counter />);
        const incrementButton = screen.getByText("Increment");

        userEvent.selectOptions(
            screen.getByRole("combobox"),
            screen.getByRole("option", { name: "5" })
        );
        userEvent.click(incrementButton);
        userEvent.click(incrementButton);
        userEvent.click(incrementButton);

        const counterText = screen.getByText("Counter: 15");

        expect(counterText).toBeInTheDocument();
    });

    it("should decrement the counter by selected step", () => {
        render(<Counter />);
        const incrementButton = screen.getByText("Increment");
        const decrementButton = screen.getByText("Decrement");

        // select increment by 5
        userEvent.selectOptions(
            screen.getByRole("combobox"),
            screen.getByRole("option", { name: "5" })
        );

        // increment 3 times
        userEvent.click(incrementButton);
        userEvent.click(incrementButton);
        userEvent.click(incrementButton);

        // decrement 2 times
        userEvent.click(decrementButton);
        userEvent.click(decrementButton);

        const counterText = screen.getByText("Counter: 5");

        expect(counterText).toBeInTheDocument();
    });

    it("should not decrement when counter is 0", () => {
        render(<Counter />);
        const decrementButton = screen.getByText("Decrement");

        userEvent.click(decrementButton);

        const counterText = screen.getByText("Counter: 0");

        expect(counterText).toBeInTheDocument();
    });

    it("should not go below zero", () => {
        render(<Counter />);
        const incrementButton = screen.getByText("Increment");
        const decrementButton = screen.getByText("Decrement");

        // increment by 1, 3 times
        userEvent.click(incrementButton);
        userEvent.click(incrementButton);
        userEvent.click(incrementButton);

        // select step 5
        userEvent.selectOptions(
            screen.getByRole("combobox"),
            screen.getByRole("option", { name: "5" })
        );
        userEvent.click(decrementButton);

        const counterText = screen.getByText("Counter: 3");

        expect(counterText).toBeInTheDocument();
    });
});
