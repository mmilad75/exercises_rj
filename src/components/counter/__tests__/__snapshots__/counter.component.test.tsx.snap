// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component should render correctly 1`] = `
<DocumentFragment>
  <div
    style="text-align: center;"
  >
    <h2>
      Counter: 0
    </h2>
    <button
      style="padding: 0.5rem 1rem; margin: 0px 0.5rem; border-radius: 4px; background: rgb(20, 23, 26); color: rgb(255, 255, 255);"
    >
      Increment
    </button>
    <button
      style="padding: 0.5rem 1rem; margin: 0px 0.5rem; border-radius: 4px; background: rgb(20, 23, 26); color: rgb(255, 255, 255);"
    >
      Decrement
    </button>
    <select
      style="padding: 0.4rem 1rem; margin: 0px 0.5rem; border-color: #14171a; border-radius: 4px;"
    >
      <option
        value="1"
      >
        1
      </option>
      <option
        value="5"
      >
        5
      </option>
      <option
        value="10"
      >
        10
      </option>
    </select>
  </div>
</DocumentFragment>
`;
